tests/ops/util/test_saas_util.py

```diff
@@ -329,50 +329,63 @@ def test_regex_is_not_greedy(self):
         )
 
 
-def test_unflatten_dict():
-    # empty dictionary
-    assert unflatten_dict({}) == {}
-
-    # empty dictionary value
-    assert unflatten_dict({"A": {}}) == {"A": {}}
+class TestUnflattenDict:
+    def test_empty_dict(self):
+        assert unflatten_dict({}) == {}
+
+    def test_empty_dict_value(self):
+        assert unflatten_dict({"A": {}}) == {"A": {}}
+
+    def test_unflattened_dict(self):
+        assert unflatten_dict({"A": "1"}) == {"A": "1"}
+
+    def test_same_level(self):
+        assert unflatten_dict({"A.B": "1", "A.C": "2"}) == {"A": {"B": "1", "C": "2"}}
+
+    def test_mixed_levels(self):
+        assert unflatten_dict(
+            {
+                "A": "1",
+                "B.C": "2",
+                "B.D": "3",
+            }
+        ) == {
+            "A": "1",
+            "B": {"C": "2", "D": "3"},
+        }
 
-    # unflattened dictionary
-    assert unflatten_dict({"A": "1"}) == {"A": "1"}
+    def test_long_path(self):
+        assert unflatten_dict({"A.B.C.D.E.F.G": "1"}) == {
+            "A": {"B": {"C": {"D": {"E": {"F": {"G": "1"}}}}}}
+        }
 
-    # same level
-    assert unflatten_dict({"A.B": "1", "A.C": "2"}) == {"A": {"B": "1", "C": "2"}}
+    def test_single_item_array(self):
+        assert unflatten_dict({"A.0.B": "C"}) == {"A": [{"B": "C"}]}
 
-    # mixed levels
-    assert unflatten_dict(
-        {
-            "A": "1",
-            "B.C": "2",
-            "B.D": "3",
+    def test_multi_item_array(self):
+        assert unflatten_dict({"A.0.B": "C", "A.1.D": "E"}) == {
+            "A": [{"B": "C"}, {"D": "E"}]
         }
-    ) == {
-        "A": "1",
-        "B": {"C": "2", "D": "3"},
-    }
-
-    # long path
-    assert unflatten_dict({"A.B.C.D.E.F.G": "1"}) == {
-        "A": {"B": {"C": {"D": {"E": {"F": {"G": "1"}}}}}}
-    }
-
-    # incoming values should overwrite existing values
-    assert unflatten_dict({"A.B": 1, "A.B": 2}) == {"A": {"B": 2}}
-
-    # conflicting types
-    with pytest.raises(FidesopsException):
-        unflatten_dict({"A.B": 1, "A": 2, "A.C": 3})
-
-    # data passed in is not completely flattened
-    with pytest.raises(FidesopsException):
-        unflatten_dict({"A.B.C": 1, "A": {"B.D": 2}})
-
-    # unflatten_dict shouldn't be called with a None separator
-    with pytest.raises(IndexError):
-        unflatten_dict({"": "1"}, separator=None)
+
+    def test_multi_value_array(self):
+        assert unflatten_dict(
+            {"A.0.B": "C", "A.0.D": "E", "A.1.F": "G", "A.1.H": "I"}
```

Comment:
i guess if we _really_ wanted to go crazy, you could test _nested lists_, i.e. lists of lists. but that feels too contrived i think 😄

Reply:
😆 agreed, let's draw the line here for now